Development/src/pages/devices/EditMatrixNamesField.js

```diff
@@ -0,0 +1,335 @@
+import React, { useEffect, useRef, useState } from 'react';
+import { IconButton, TextField, Typography } from '@material-ui/core';
+import CreateIcon from '@material-ui/icons/Create';
+import DeleteIcon from '@material-ui/icons/Delete';
+import ClearIcon from '@material-ui/icons/Clear';
+import DoneIcon from '@material-ui/icons/Done';
+import { get, has, isEmpty, set } from 'lodash';
+
+export const EditableIONameField = ({
+    defaultValue,
+    source,
+    label,
+    customNames,
+    setCustomNames,
```

Comment:
I think the idiomatic approach to reading and writing the custom names would be to use the [`useContext` hook](https://reactjs.org/docs/hooks-reference.html#usecontext). After a quick play, I think this would save 100+ lines of passing things through the nested components to where they are used. Something to look at later...

Reply:
Both the custom names state and the tooltip modal state are now managed using context, provided by the ChannelMappingMatrix component itself and consumed only where necessary.